### ISSUE TITLE
Move function spaces to configuration file 

### ISSUE THREAD

**Issue by TeranIvy** [stfc/PSyclone#742]
Currently the function spaces in LFRic API are hardcoded into `dynamo0p3.py`. It would be more flexible to move lists of continuous and discontinuous function spaces into the configuration file so they can be added without having to change LFRic API in PSyclone.

Whilst this would be much more flexible than the current approach, it also raises a series of questions.

1. Formulas for number of DoFs per cell that can be used for kernel transformations (see `class Dynamo0p3KernelConstTrans(Transformation)` in `transformations.py`). We need to add them for each new supported function space. Where would they be stored now? My suggestion would be to add the formula when adding a new function space in the configuration file in the form of e.g.
  ```
  discontinuous_function_spaces = w3: (n+1)**3, wtheta: (n+2)*(n+1)**2, ...
  ```
 Also, how do we test them in PSyclone if they are added just to the configuration file?

2. How do we ensure that the function spaces added in the config file continuous and discontinuous lists are indeed continuous or discontinuous if the PSyclone tests are not updated?

3. How do we ensure that we generate correct code for kernel calls and stubs in PSyclone if the function spaces in config file are updated and/or reordered and PSyclone tests are not between the releases?

4. Each new function space is described in `dynamo0p3.rst` user guide. If the function spaces are added only in config file, how would we update PSyclone documentation?

I would appreciate thoughts from @rupertford, @arporter, @thomasmelvin and @christophermaynard about these issues.

**Comment by TeranIvy** [comment]
Added @stevemullerworth and @MatthewHambley for the question of governance and testing.

**Comment by thomasmelvin** [comment]
We are only adding new function spaces slowly (and less frequently as time goes on), so I'm not sure it's worth spending lots of effort on this in the short term to make the May deadline unless being more configurable gives use lots of other benefits?

The new spaces we need for the hybridization preconditioner are the W2v_trace and W2h_trace and these are implemented in [lfric:#2149](https://code.metoffice.gov.uk/trac/lfric/ticket/2149). In terms of continuity they are exactly the same as W2v and W2h (continuous through vertical and horizontal faces) and the dof numbers are similar, they have the sanme number of dofs on faces, but dont have any dofs in the cell volumes, i..e for order p elements:
W2v_trace has `(number of vertical faces=2)*(p+1)*(p+1)` dofs and 
W2h_trace has `(number of horizontal faces=4)*(p+1)*(p+1)` dofs

**Comment by MatthewHambley** [comment]
> Formulas for number of DoFs per cell that can be used for kernel transformations (see `class Dynamo0p3KernelConstTrans(Transformation)` in `transformations.py`). We need to add them for each new supported function space. Where would they be stored now? My suggestion would be to add the formula when adding a new function space in the configuration file in the form of e.g.
> 
> ```
> discontinuous_function_spaces = w3: (n+1)**3, wtheta: (n+2)*(n+1)**2, ...
> ```

Is this the right way round to consider this problem? Would it be better to consider the functionspace as the primary concept with properties thereof. i.e.

```
function_spaces = {'w3': FunctionSpace(dofs_per_cell='(n+1)**3', horizontally_continuous=False),
                   'wtheta': FunctionSpace(dofs_per_cell='(n+2)*(n+1)**2', horizontally_continuous=False),
                   'w2': FunctionSpace(dofs_per_cell='...', horizontally_continuous=True)}
```

That way it is easy to add additional properties without upsetting anything else. It also means that once you have enquired about a function space you know everything there is to know about it. Rather than having to enquire of a number of different dictionaries.

> How do we ensure that we generate correct code for kernel calls and stubs in PSyclone if the function spaces in config file are updated and/or reordered and PSyclone tests are not between the releases?

Would it be helpful to test head-of-PSyclone against head-of-LFRic on a regular basis or is that going to become quickly annoying?

> Each new function space is described in `dynamo0p3.rst` user guide. If the function spaces are added only in config file, how would we update PSyclone documentation?

The configuration file can be, to a certain extent, self documenting. There should probably be documentation on the LFRic side which describes all our function spaces.

> On the LFRic side, who has the right to update the PSyclone configuration file with the new function spaces (and through which procedure - user request or something else) and how do we ensure that we have uniformity over various platforms that we use?

While we absolutely want to encourage people to specify new functionspaces tailored to their need rather than bodging with existing ones we also want to prevent unfettered proliferation. If this configuration file lives in the LFRic repository then it may be a simple matter of adding its governance to code review or design compliance review.

**Comment by stevemullerworth** [comment]
Some early thoughts were recorded here which I think partly answer some questions.
https://code.metoffice.gov.uk/trac/lfric/ticket/414

In the past we have talked about a generic way of defining function spaces that both LFRic and PSyclone could use. Might there be benefits to having information about entities that dofs live on (if that can be part of the configuration). For example:

1. The information can be used to infer continuity (horizontally or vertically) without it being explicitly defined. Would that help with function spaces whose continuity is a function of order?
2. To transform kernels, do we need more than just dofs per cell. For example, to compute the number of dofs in a column one needs to know whether dofs are on faces between cells on different levels.

**Comment by TeranIvy** [comment]
LFRicians are currently organising infrastructure tickets and [414](https://code.metoffice.gov.uk/trac/lfric/ticket/414) was one of PSyclone-related tickets appearing in the list. As it is effectively a duplicate of this issue, I will close it. For the record, @rupertford's comments from 414 (made on 25/11/2015) are below:

>  Psyclone is mostly agnostic about function spaces and should continue to work if the list of function space names that is currently uses was increased.
>  
>  The main issues I can think of from a PSyclone correctness point of view are
>  
>  1) We would like to have a test for all supported function spaces - just in case. It could be argued that this is overkill as it does not increase the coverage of the tests if there are no special cases for a particular function space or set of function spaces
>  
>  2) We envisage having PSyclone documentation that states the function spaces that are supported and explains what they mean (or perhaps we could reference a separate document?)
>  
>  3) It would be nice to have PSyclone examples which make use of all of the function spaces (perhaps this is unrealistic).
>  
>  4) If the list is externally controlled and we have tests, documentation and examples, the external list could, in theory have function spaces removed. This would lead to an inconsistency.
>  
>  5) The PSyclone stub generator needs to know the dimensions of the basis function and differential basis function arrays. These are dependent on the function space
>  
>  6) Certain optimisation transformations could apply to certain function space types or be constrained by certain function space types. The only example we have at the moment is colouring, where colouring is is required if the space is not W3 (actually all discontinuous spaces as was pointed out) and the access is INC. Extending the idea of colouring, we can colour in different ways depending on the space and it would be useful to know which spaces are coloured in the same way, or could be coloured in the same way, so that we can check the validity of loop fusing.

**Comment by TeranIvy** [comment]
> 
> We are only adding new function spaces slowly (and less frequently as time goes on), so I'm not sure it's worth spending lots of effort on this in the short term to make the May deadline unless being more configurable gives use lots of other benefits?
> 
Agreed and created #751 to implement `W2Htrace` and `W2Vtrace` for May.
